File: android/src/main/java/app/capgo/barometer/CapacitorBarometerPlugin.java
package app.capgo.barometer;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.SystemClock;
import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "CapacitorBarometer")
public class CapacitorBarometerPlugin extends Plugin implements SensorEventListener {

    private final String pluginVersion = "8.0.11";
    private static final String PERMISSION_GRANTED = "granted";
    private static final String PERMISSION_DENIED = "denied";

    @Nullable
    private SensorManager sensorManager;

    @Nullable
    private Sensor barometer;

    private final JSObject lastMeasurement = new JSObject();

    private boolean updatesActive = false;

    @Override
    public void load() {
        Context context = getContext();
        sensorManager = (SensorManager) context.getSystemService(Context.SENSOR_SERVICE);
        if (sensorManager != null) {
            barometer = sensorManager.getDefaultSensor(Sensor.TYPE_PRESSURE);
        }
        lastMeasurement.put("pressure", 0);
        lastMeasurement.put("relativeAltitude", 0);
        lastMeasurement.put("timestamp", 0.0);
    }

    @PluginMethod
    public void getMeasurement(PluginCall call) {
        if (barometer == null) {
            call.reject("Barometer sensor not available on this device.");
            return;
        }
        call.resolve(snapshotMeasurement());
    }

    @PluginMethod
    public void isAvailable(PluginCall call) {
        JSObject result = new JSObject();
        result.put("isAvailable", barometer != null);
        call.resolve(result);
    }

    @PluginMethod
    public void startMeasurementUpdates(PluginCall call) {
        if (barometer == null) {
            call.reject("Barometer sensor not available on this device.");
            return;
        }
        if (updatesActive) {
            call.resolve();
            return;
        }
        if (sensorManager != null && sensorManager.registerListener(this, barometer, SensorManager.SENSOR_DELAY_NORMAL)) {
            updatesActive = true;
            call.resolve();
        } else {
            call.reject("Failed to register barometer listener.");
        }
    }

    @PluginMethod
    public void stopMeasurementUpdates(PluginCall call) {
        if (sensorManager != null && updatesActive) {
            sensorManager.unregisterListener(this);
            updatesActive = false;
        }
        call.resolve();
    }

    @PluginMethod
    public void checkPermissions(PluginCall call) {
        JSObject result = new JSObject();
        result.put("barometer", barometer != null ? PERMISSION_GRANTED : PERMISSION_DENIED);
        call.resolve(result);
    }

    @PluginMethod
    public void requestPermissions(PluginCall call) {
        checkPermissions(call);
    }

    @PluginMethod
    public void removeAllListeners(PluginCall call) {
        super.removeAllListeners(call);
    }

    @Override
    public void handleOnDestroy() {
        if (sensorManager != null) {
            sensorManager.unregisterListener(this);
        }
        super.handleOnDestroy();
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        if (event.sensor.getType() != Sensor.TYPE_PRESSURE) {
            return;
        }

        double pressure = event.values[0]; // Already reported in hPa
        // SensorEvent.timestamp is nanoseconds since boot (elapsedRealtimeNanos
        // clock). Convert to Unix epoch by computing the sensor sample's age
        // relative to the current boot clock and subtracting from wall clock.
        // This preserves the sensor's true measurement time instead of using
        // System.currentTimeMillis() which reflects callback delivery time.
        long sensorAgeNanos = SystemClock.elapsedRealtimeNanos() - event.timestamp;
        double timestamp = System.currentTimeMillis() - sensorAgeNanos / 1_000_000.0;

        lastMeasurement.put("pressure", pressure);
        lastMeasurement.put("relativeAltitude", 0);
        lastMeasurement.put("timestamp", timestamp);

        JSObject payload = new JSObject();
        payload.put("pressure", pressure);
        payload.put("relativeAltitude", 0);
        payload.put("timestamp", timestamp);
        notifyListeners("measurement", payload);
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
        // No-op
    }

    private JSObject snapshotMeasurement() {
        JSObject copy = new JSObject();
        copy.put("pressure", lastMeasurement.optDouble("pressure", 0));
        copy.put("relativeAltitude", lastMeasurement.optDouble("relativeAltitude", 0));
        copy.put("timestamp", lastMeasurement.optDouble("timestamp", 0));
        return copy;
    }

    @PluginMethod
    public void getPluginVersion(final PluginCall call) {
        try {
            final JSObject ret = new JSObject();
            ret.put("version", this.pluginVersion);
            call.resolve(ret);
        } catch (final Exception e) {
            call.reject("Could not get plugin version", e);
        }
    }
}

File: ios/Sources/CapacitorBarometerPlugin/CapacitorBarometerPlugin.swift
import Capacitor
import CoreMotion
import Foundation

@objc(CapacitorBarometerPlugin)
public class CapacitorBarometerPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion: String = "8.0.11"
    public let identifier = "CapacitorBarometerPlugin"
    public let jsName = "CapacitorBarometer"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getMeasurement", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startMeasurementUpdates", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopMeasurementUpdates", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeAllListeners", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let altimeter = CMAltimeter()
    private var updatesActive = false
    private var latestMeasurement: [String: Any] = [
        "pressure": 0.0,
        "relativeAltitude": 0.0,
        "timestamp": 0.0
    ]

    @objc func getMeasurement(_ call: CAPPluginCall) {
        guard CMAltimeter.isRelativeAltitudeAvailable() else {
            call.reject("Barometer not available")
            return
        }

        call.resolve(latestMeasurement)
    }

    @objc func isAvailable(_ call: CAPPluginCall) {
        call.resolve(["isAvailable": CMAltimeter.isRelativeAltitudeAvailable()])
    }

    @objc func startMeasurementUpdates(_ call: CAPPluginCall) {
        guard CMAltimeter.isRelativeAltitudeAvailable() else {
            call.reject("Barometer not available")
            return
        }

        if updatesActive {
            call.resolve()
            return
        }

        altimeter.startRelativeAltitudeUpdates(to: OperationQueue.main) { [weak self] data, error in
            guard let self else { return }

            if let error {
                CAPLog.print("CapacitorBarometerPlugin", "Barometer error: \(error.localizedDescription)")
                return
            }

            guard let data else { return }

            let pressureHectoPascal = data.pressure.doubleValue * 10.0
            let relativeAltitudeMeters = data.relativeAltitude.doubleValue
            // CMAltimeterData.timestamp is seconds since device boot (system uptime).
            // Convert to Unix epoch by computing the age of the sample relative to
            // the current uptime and subtracting that from the current wall clock.
            // This preserves the sensor's true measurement time instead of using
            // Date() which reflects when the callback fires — a critical difference
            // when iOS delivers batched updates after backgrounding.
            let sensorAge = ProcessInfo.processInfo.systemUptime - data.timestamp
            let timestamp = (Date().timeIntervalSince1970 - sensorAge) * 1000

            let measurement: [String: Any] = [
                "pressure": pressureHectoPascal,
                "relativeAltitude": relativeAltitudeMeters,
                "timestamp": timestamp
            ]

            self.latestMeasurement = measurement
            DispatchQueue.main.async {
                self.notifyListeners("measurement", data: measurement)
            }
        }

        updatesActive = true
        call.resolve()
    }

    @objc func stopMeasurementUpdates(_ call: CAPPluginCall) {
        if updatesActive {
            altimeter.stopRelativeAltitudeUpdates()
            updatesActive = false
        }
        call.resolve()
    }

    @objc override public func checkPermissions(_ call: CAPPluginCall) {
        call.resolve(["barometer": currentPermissionState()])
    }

    @objc override public func requestPermissions(_ call: CAPPluginCall) {
        call.resolve(["barometer": currentPermissionState()])
    }

    @objc override public func removeAllListeners(_ call: CAPPluginCall) {
        super.removeAllListeners(call)
    }

    private func currentPermissionState() -> String {
        guard CMAltimeter.isRelativeAltitudeAvailable() else {
            return "denied"
        }

        if #available(iOS 17.0, *) {
            switch CMAltimeter.authorizationStatus() {
            case .authorized:
                return "granted"
            case .denied, .restricted:
                return "denied"
            case .notDetermined:
                return "prompt"
            @unknown default:
                return "prompt"
            }
        }

        return "granted"
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": self.pluginVersion])
    }
}
